common/static/coffee/spec/discussion/discussion_router_spec.coffee

```diff
@@ -0,0 +1,189 @@
+        comments_count: 1,
+        created_at: '2013-04-03T20:08:39Z',
+      }),
+      DiscussionViewSpecHelper.makeThreadWithProps({
+        id: "2",
+        title: "Thread2",
+        votes: {up_count: '42'},
+        comments_count: 2,
+        created_at: '2013-04-03T20:07:39Z',
+      }),
+      DiscussionViewSpecHelper.makeThreadWithProps({
+        id: "3",
+        title: "Thread3",
+        votes: {up_count: '12'},
+        comments_count: 3,
+        created_at: '2013-04-03T20:06:39Z',
+      }),
+      DiscussionViewSpecHelper.makeThreadWithProps({
+        id: "4",
+        title: "Thread4",
+        votes: {up_count: '25'},
+        comments_count: 0,
+        pinned: true,
+        created_at: '2013-04-03T20:05:39Z',
+      }),
+    ]
+
+    spyOn(DiscussionUtil, 'makeWmdEditor')
+    @discussion = new Discussion(_.map(@threads, (thread_spec) -> new Thread(thread_spec)), {pages: 2, sort: 'date'})
+    @course_settings = new DiscussionCourseSettings({
+          "category_map": {
+            "children": ["Topic", "General"],
+            "entries": {
+              "Topic": {"is_cohorted": false, "id": "topic"},
+              "General": {"is_cohorted": false, "id": "general"}
+            }
+          },
+          "allow_anonymous": false,
+          "allow_anonymous_to_peers": false,
+          "is_cohorted": true,
+          "cohorts": [
+            {"id": 1, "name": "Cohort1"},
+            {"id": 2, "name": "Cohort2"}
+          ]
+        })
+    @router = new DiscussionRouter({discussion: @discussion, course_settings: @course_settings})
+
+  describe 'showThread', ->
+    existingCheck = (thread_id) ->
+      it "shows thread #{thread_id}, which is already in threads collection", ->
+        DiscussionSpecHelper.makeAjaxSpy(() -> )
+        spyOn(@router, 'renderThreadView')
+        # precondition check - thread is in router's collection
+        expect(@router.discussion.get(thread_id)).not.toBeUndefined()
+        @router.showThread("irrelevant forum name", thread_id)
+        expect($.ajax).not.toHaveBeenCalled()
+        expect(@router.renderThreadView).toHaveBeenCalled()
+
+    missingCheck = (forum_name, thread_id) ->
+      it "requests thread #{thread_id} in forum #{forum_name} if not already in collection", ->
+
+        DiscussionSpecHelper.makeAjaxSpy(
+          (params) -> expect(params.url.path()).toBe(DiscussionUtil.urlFor('retrieve_single_thread', forum_name, thread_id))
+        )
+        spyOn(@router, 'renderThreadView')
+        # precondition check - thread is in router's collection
+        expect(@router.discussion.get(thread_id)).toBeUndefined()
+        @router.showThread(forum_name, thread_id)
+        # in this case showThread makes a hidden async ajax request - it's hard to hook inot it, so it's simpler and
+        # sufficient to just schedule assertion to run as soon as possible, but not immediately
+        setTimeout (() -> expect(@router.renderThreadView).toHaveBeenCalled()), 0
```

Comment:
This doesn't work correctly. The spec finishes running before the scheduled timeout gets executed and the expectation in the timeout is ignored. You can replace `.toHaveBeenCalled()` with `.not.toHaveBeenCalled()` and the spec will still pass.
Jasmine does support async specs, but you need to defined them a bit differently (see http://jasmine.github.io/2.0/introduction.html#section-Asynchronous_Support for Jasmine 2, not sure what version is used here).

The problem with this spec is not that it's async though, it's that you stub out `$.ajax` (with `makeAjaxSpy`) and ajax callbacks are never executed.

You should make sure that the `success` ajax callback gets executed. There are probably prettier ways to do it, but instead of using `makeAjaxSpy`, something like this should work:

```coffeescript
spyOn($, 'ajax').andCallFake(
    expect(params.url.path()).toBe(DiscussionUtil.urlFor('retrieve_single_thread', forum_name, thread_id))
    d = $.Deferred()
    d.done(params.success)
    d.resolveWith(null, [{content: 'whatever'}])
)
```

This way the spec remains synchronous so you don't have to use timeouts and you can control the data that gets passed to `renderThreadView`.